voip_utils/call_phone.py

```diff
@@ -0,0 +1,229 @@
+        self.call_protocol_factory = call_protocol_factory
+        self._tasks: Set[asyncio.Future[Any]] = set()
+
+    def on_call(self, call_info: CallInfo):
+        """Answer incoming calls and start RTP server on a random port."""
+
+        rtp_ip = self._source_endpoint.host
+
+        _LOGGER.debug(
+            "Starting RTP server on ip=%s, rtp_port=%s, rtcp_port=%s",
+            rtp_ip,
+            self._rtp_port,
+            self._rtp_port + 1,
+        )
+
+        # Handle RTP packets in RTP server
+        rtp_task = asyncio.create_task(
+            self._create_rtp_server(
+                self.call_protocol_factory, call_info, rtp_ip, self._rtp_port
+            )
+        )
+        self._tasks.add(rtp_task)
+        rtp_task.add_done_callback(self._tasks.remove)
+
+        _LOGGER.debug("RTP server started")
+
+    def end_call(self, task):
+        """Callback for hanging up when call is ended."""
+        self.hang_up()
+
+    async def _create_rtp_server(
+        self,
+        protocol_factory: CallProtocolFactory,
+        call_info: CallInfo,
+        rtp_ip: str,
+        rtp_port: int,
+    ):
+        # Shared state between RTP/RTCP servers
+        rtcp_state = RtcpState()
+
+        loop = asyncio.get_running_loop()
+
+        # RTCP server
+        await loop.create_datagram_endpoint(
+            lambda: RtcpDatagramProtocol(rtcp_state),
+            (rtp_ip, rtp_port + 1),
+        )
+
+        # RTP server
+        await loop.create_datagram_endpoint(
+            partial(protocol_factory, call_info, rtcp_state),
+            (rtp_ip, rtp_port),
+        )
+
+
+class PreRecordMessageProtocol(RtpDatagramProtocol):
+    """Plays a pre-recorded message on a loop."""
+
+    def __init__(
+        self,
+        file_name: str,
+        opus_payload_type: int,
+        message_delay: float = 1.0,
+        loop_delay: float = 2.0,
+        loop: Optional[asyncio.AbstractEventLoop] = None,
+        rtcp_state: RtcpState | None = None,
+    ) -> None:
+        """Set up RTP server."""
+        super().__init__(
+            rate=RATE,
+            width=WIDTH,
+            channels=CHANNELS,
+            opus_payload_type=opus_payload_type,
+            rtcp_state=rtcp_state,
+        )
+        self.loop = loop
+        self.file_name = file_name
+        self.message_delay = message_delay
+        self.loop_delay = loop_delay
+        self._audio_task: asyncio.Task | None = None
+        self._audio_bytes: bytes | None = None
+        _LOGGER.debug("Created PreRecordMessageProtocol")
+
+    def on_chunk(self, audio_bytes: bytes) -> None:
+        """Handle raw audio chunk."""
+        _LOGGER.debug("on_chunk")
+        if self.transport is None:
+            return
+
+        if self._audio_bytes is None:
+            # 16Khz, 16-bit mono audio message
+            file_path = Path(__file__).parent / self.file_name
+            self._audio_bytes = file_path.read_bytes()
```

Comment:
Note: This is doing I/O in the event loop. That's fine as this is a demo protocol and any usage would have the app pass in the bytes anyway.